qCC/ccSubsamplingDlg.cpp

```diff
@@ -66,109 +69,153 @@ ccSubsamplingDlg::~ccSubsamplingDlg()
 	delete m_ui;
 }
 
-CCCoreLib::ReferenceCloud* ccSubsamplingDlg::getSampledCloud(ccGenericPointCloud* cloud, CCCoreLib::GenericProgressCallback* progressCb/*=nullptr*/)
-{
+CCCoreLib::ReferenceCloud* ccSubsamplingDlg::getSampledCloud(ccGenericPointCloud* cloud, CCCoreLib::GenericProgressCallback* progressCb/*=nullptr*/) const {
+	CCCoreLib::ReferenceCloud *result = nullptr;
+
+	try {
+		result = DoSubSample(subSamplingParams(), cloud, progressCb);
+	} catch (std::exception& e) {
+		ccLog::Warning(QString("[ccSubsamplingDlg::getSampledCloud] Failed to subsample %1: %2!").arg(cloud->getName(), e.what()));
+	}
+	return result;
+}
+
+
+bool SubsamplingParams::Spatial::tryEnableSfModulation(const ccScalarField& sf) {
+	sfMin = CCCoreLib::NAN_VALUE;
+	sfMax = CCCoreLib::NAN_VALUE;
+	if (!ccScalarField::ValidValue(sfMin) || sfMin > sf.getMin())
+		sfMin = sf.getMin();
+	if (!ccScalarField::ValidValue(sfMax) || sfMax < sf.getMax())
+		sfMax = sf.getMax();
+
+	if (!ccScalarField::ValidValue(sfMin) || !ccScalarField::ValidValue(sfMax))
+	{
+		// warn the user, don't use 'Use Active SF' and keep going
+		ccLog::Warning(QObject::tr("\tCan't use 'scalar field modulation': scalar field '%1' has invalid min/max values.").arg(QString::fromStdString(sf.getName())));
+		modulationEnabled = false;
+	} else {
+		modulationEnabled = true;
+	}
+	return modulationEnabled;
+}
+
+
+CCCoreLib::ReferenceCloud* DoSubSample(const SubsamplingParams &params, ccGenericPointCloud *cloud, CCCoreLib::GenericProgressCallback *progressCb /*=nullptr*/) {
 	if (!cloud || cloud->size() == 0)
 	{
-		ccLog::Warning("[ccSubsamplingDlg::getSampledCloud] Invalid input cloud!");
-		return nullptr;
+		throw std::runtime_error("Invalid input cloud!");
```

Comment:
In CloudCompare, I prefer not too propagate exceptions, and manage errors as soon as they happen.

Reply:
I expected a comment like that :grin: 

I'm starting to  think that to do this we may need to introduce some kind "UserFeedback" interface (and have CmdUserFeeback and GUIUserFeeback implementations), as probably, for warning ccWarning is fine, but for errors we probably want to display differently between GUI and cmd (and maybe later python scripts)

Reply:
Or maybe simply add some static parameters/methods to change the behavior of ccLog::Error in command line mode? (or Python)

Reply:
I'm confused with how logging is handled between regular GUI mode, cmdline non-silent, cmdline silent

I see that commandline interface has its own methods for logging, that the commands use as `cmd.print`, `cmd.error`. 

To me only `ccLog::` should exists and be used, and then the actual instance  would handle GUI vs CMD mode non-SILENT vs CMD mode SILENT, either by having different subclasses for ccLog or ccConsole handling those via some settings

Looking at the code, the cmd.error and co seems to forward to ccConsole, so it seems things are correctly made (except I could not find where the message box for Error level is disabled in silent mode) but the cmd.print family of functions do not really provide that much value and at least for me are creating some confusion

Reply:
Is this something we should try to adress little by little ?

Reply:
Ah yes, don't hesitate!